File: python/say.py
NUMBER_MAP = {
    0: '',
    1: 'one',
    2: 'two',
    3: 'three',
    4: 'four',
    5: 'five',
    6: 'six',
    7: 'seven',
    8: 'eight',
    9: 'nine',
    10: 'ten',
    11: 'eleven',
    12: 'twelve',
    13: 'thirteen',
    14: 'fourteen',
    15: 'fifteen',
    16: 'sixteen',
    17: 'seventeen',
    18: 'eighteen',
    19: 'nineteen',
    20: 'twenty',
    30: 'thirty',
    40: 'forty',
    50: 'fifty',
    60: 'sixty',
    70: 'seventy',
    80: 'eighty',
    90: 'ninety',
}


def aux(number):
    div = number // 100
    number = number % 100
    said = []
    if div > 0:
        said.append(f'{NUMBER_MAP[div]} hundred')

    if number == 0:
        return ''.join(said)

    if number in NUMBER_MAP:
        said.append(f'{NUMBER_MAP[number]}')
    else:
        rem = number % 10
        number = number - rem
        if rem == 0:
            said.append(f'{NUMBER_MAP[number]}')
        else:
            said.append(f'{NUMBER_MAP[number]}-{NUMBER_MAP[rem]}')

    return ' '.join(said)


def say(number):
    if number < 0 or number > 999_999_999_999:
        raise ValueError('input out of range')

    if number == 0:
        return 'zero'

    said = []
    div = number // 1000000000
    number = number % 1000000000

    if div > 0:
        said.append(f'{aux(div)} billion')

    div = number // 1000000
    number = number % 1000000
    if div > 0:
        said.append(f'{aux(div)} million')

    div = number // 1000
    number = number % 1000
    if div > 0:
        said.append(f'{aux(div)} thousand')

    said.append(f'{aux(number)}')

    return ' '.join([r for r in said if r != ''])

File: elixir/say.ex
defmodule Say do
  @doc """
  Translate a positive integer into English.
  """
  @spec in_english(integer) :: {atom, String.t()}
  def in_english(number) when number < 0 or number >= 1_000_000_000_000 do
    {:error, "number is out of range"}
  end

  def in_english(0), do: {:ok, "zero"}

  def in_english(number) do
    Stream.unfold(number, fn
      0 -> nil
      n -> {rem(n, 1000), div(n, 1000)}
    end)
    |> Stream.map(&to/1)
    |> Stream.map(&String.trim/1)
    |> Enum.zip(["", "thousand", "million", "billion"])
    |> Enum.map(fn
      {"", _} -> ""
      {s, ""} -> "#{s}"
      {s, t} -> "#{s} #{t}"
    end)
    |> Enum.reject(&(&1 == ""))
    |> Enum.reverse()
    |> Enum.join(" ")
    |> then(&{:ok, &1})
  end

  def to(0), do: ""
  def to(1), do: "one"
  def to(2), do: "two"
  def to(3), do: "three"
  def to(4), do: "four"
  def to(5), do: "five"
  def to(6), do: "six"
  def to(7), do: "seven"
  def to(8), do: "eight"
  def to(9), do: "nine"
  def to(10), do: "ten"
  def to(11), do: "eleven"
  def to(12), do: "twelve"
  def to(13), do: "thirteen"
  def to(14), do: "fourteen"
  def to(15), do: "fifteen"
  def to(16), do: "sixteen"
  def to(17), do: "seventeen"
  def to(18), do: "eighteen"
  def to(19), do: "nineteen"
  def to(20), do: "twenty"
  def to(30), do: "thirty"
  def to(40), do: "forty"
  def to(50), do: "fifty"
  def to(60), do: "sixty"
  def to(70), do: "seventy"
  def to(80), do: "eighty"
  def to(90), do: "ninety"

  def to(n) when n < 100 do
    "#{div(n, 10) |> Kernel.*(10) |> to()}-#{rem(n, 10) |> to()}"
  end

  def to(n) do
    "#{div(n, 100) |> to()} hundred #{rem(n, 100) |> to()}"
  end
end
